### PR TITLE
DT-423: Addition of MFA route

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaController.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  @GetMapping
+  open fun mfaChallengeRequest(@RequestParam token: String,
+                               request: HttpServletRequest, response: HttpServletResponse): ModelAndView? {
+
+    val optionalErrorCode = tokenService.checkToken(TokenType.MFA, token)
+    if (optionalErrorCode.isPresent) {
+      return ModelAndView("redirect:/login?error=mfa${optionalErrorCode.get()}")
+    }
+
+//    return ModelAndView("mfaChallenge", "token", token)
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/MfaController.kt]: temporarily bypass the mfa page altogether as changes were getting too big

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserToken.java [hunk 2]
```diff
@@ -36,7 +36,7 @@
         this.user = user;
 
         final var now = LocalDateTime.now();
-        this.tokenExpiry = tokenType == TokenType.CHANGE ? now.plusMinutes(20) : now.plusDays(1);
+        this.tokenExpiry = tokenType == TokenType.CHANGE || tokenType == TokenType.MFA ? now.plusMinutes(20) : now.plusDays(1);
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserToken.java]: will try 20 minutes initially for MFA and see how we get on

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      is CredentialsExpiredException -> {
+        // special handling for expired users and feature switch turned on
+        val username = StringUtils.trim(request.getParameter("username").toUpperCase())
+        val token = tokenService.createToken(TokenType.CHANGE, username)
+        redirectStrategy.sendRedirect(request, response, "/change-password?token=$token")
+        return
+      }
+      is MfaRequiredException -> {
+        // need to break out to perform mfa for the user
+        val username = StringUtils.trim(request.getParameter("username").toUpperCase())
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/security/UserStateAuthenticationFailureHandler.kt]: add in mfa challenge

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -95,6 +95,9 @@ application:
...
     match-subdomains: false
+    mfa:
+      whitelist: 172.16.0.0/12
+      roles: ROLE_MFA
```
Comment (petergphillips) [inline src/main/resources/application.yml]: For initial attempt, user needs a `MFA` role in order to trigger MFA

File: src/test/groovy/uk/gov/justice/digital/hmpps/oauth2server/integration/specs/MfaSpecification.groovy [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  def "Login as user with MFA enabled"() {
+    given: 'I try to login with a user with MFA enabled'
+    to LoginPage
+
+    when: 'I login'
+    loginAs AUTH_MFA_USER, 'password123456'
+
+//        and: 'I am redirected to the mfa page'
```
Comment (petergphillips) [inline src/test/groovy/uk/gov/justice/digital/hmpps/oauth2server/integration/specs/MfaSpecification.groovy]: commented out as mfa bypasses the challenge page for now

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserToken.java [hunk 6]
```diff
@@ -63,7 +63,9 @@ public User getUser() {
     public enum TokenType {
         RESET("ResetPassword"),
         VERIFIED("VerifiedPassword"),
-        CHANGE("ChangePassword");
+        CHANGE("ChangePassword"),
+        MFA("MFA"),
+        MFA_CODE("MFACode");
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/auth/model/UserToken.java]: need two token types - one to grant the user access the MFA challenge page and subsequent login and the other for the MFA code that gets emailed to them

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/service/MfaServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  @After
+  fun tearDown() {
+    RequestContextHolder.resetRequestAttributes()
+  }
+
+  @Test
+  fun `needsMfa whitelisted IP`() {
+    request.remoteAddr = "12.21.23.24"
+
+    assertThat(service.needsMfa(emptySet())).isFalse()
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/digital/hmpps/oauth2server/service/MfaServiceTest.kt]: shouldn't this be `isTrue()` ? I'm a bit confused

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import uk.gov.justice.digital.hmpps.oauth2server.utils.IpAddressHelper
+import uk.gov.justice.digital.hmpps.oauth2server.verify.TokenService
+import java.util.*
+
+@Service
+open class MfaService(@Value("\${application.authentication.mfa.whitelist}") whitelist: Set<String>,
+                      @Value("\${application.authentication.mfa.roles}") roles: Set<String>,
+                      private val tokenService: TokenService) {
+  private val ipMatchers: List<IpAddressMatcher>
+  private val roles: Set<String>
```
[the remark this answers is not in the document]
Comment (petergphillips) [reply src/main/java/uk/gov/justice/digital/hmpps/oauth2server/service/MfaService.kt]: :+1: will do on next commit

**Pull request merged**